src/affine_dimensions.jl

```diff
@@ -0,0 +1,390 @@
+AffineDimensions(d::Dimensions{R}) where R = AffineDimensions{R}(scale=1.0, offset=0.0, basedim=d, symbol=:nothing)
+
+# Handle offsets in affine dimensions
+function AffineDimensions{R}(s::Real, o::Real, dims::AbstractAffineDimensions, sym=:nothing) where {R}
+    new_s = s * affine_scale(dims)
+    new_o = affine_offset(dims) + o * affine_scale(dims)
+    return AffineDimensions{R}(new_s, new_o, affine_base_dim(dims), sym)
+end
+
+function AffineDimensions{R}(s::Real, o::UnionAbstractQuantity, dims::AbstractAffineDimensions, sym=:nothing) where {R}
+    new_s = s * affine_scale(dims)
+    new_o = affine_offset(dims) + ustrip(uexpand(o))
+    return AffineDimensions{R}(new_s, new_o, affine_base_dim(dims), sym)
+end
+
+function AffineDimensions{R}(s::Real, o::UnionAbstractQuantity, dims::Dimensions, sym=:nothing) where {R}
+    return AffineDimensions{R}(s, ustrip(uexpand(o)), dims, sym)
+end
+
+# From two quantities 
+function AffineDimensions{R}(s::Real, o::UnionAbstractQuantity, q::UnionAbstractQuantity, sym=:nothing) where {R}
+    q_si_origin = uexpand(0 * q)
+    o_si_origin = uexpand(0 * o)
+    o_difference_to_si = uexpand(o) - o_si_origin
+    dimension(q_si_origin) == dimension(o_difference_to_si) || throw(DimensionError(o, q))
+    o_si = o_difference_to_si + q_si_origin
+    q_si = uexpand(q) - q_si_origin
+    return AffineDimensions{R}(s, o_si, q_si, sym)
+end
+
+# Base case with SI units
+function AffineDimensions{R}(s::Real, o::UnionAbstractQuantity{<:Any,<:Dimensions}, q::UnionAbstractQuantity{<:Any,<:Dimensions}, sym=:nothing) where {R}
+    dimension(o) == dimension(q) || throw(DimensionError(o, q))
+    return AffineDimensions{R}(s * ustrip(q), ustrip(o), dimension(q), sym)
+end
+
+# Offset from real
+function AffineDimensions{R}(s::Real, o::Real, q::Q, sym=:nothing) where {R, Q<:UnionAbstractQuantity}
+    return AffineDimensions{R}(s, o * q, q, sym)
+end
+
+affine_scale(d::AffineDimensions) = d.scale
+affine_offset(d::AffineDimensions) = d.offset
+affine_base_dim(d::AffineDimensions) = d.basedim
+
+with_type_parameters(::Type{<:AffineDimensions}, ::Type{R}) where {R} = AffineDimensions{R}
+@unstable constructorof(::Type{<:AffineDimensions}) = AffineDimensions
+
+function Base.show(io::IO, d::AbstractAffineDimensions)
+    if d.symbol != :nothing
+        print(io, d.symbol)
+    elseif isone(affine_scale(d)) && iszero(affine_offset(d))
+        print(io, affine_base_dim(d))
+    elseif iszero(affine_offset(d))
+        print(io, "(", affine_scale(d), " ", affine_base_dim(d),")")
+    else
+        addsign = affine_offset(d) < 0 ? "-" : "+"
+        if isone(affine_scale(d))
+            print(io, "(", addsign, abs(affine_offset(d)), affine_base_dim(d), ")")
+        else
+            print(io, "(", affine_scale(d), addsign, abs(affine_offset(d)), " ", affine_base_dim(d),")")
+        end
+    end
+end
```

Comment:
The printing for generic affine quantities is a bit confusing:

```julia
julia> x = AffineDimensions(1.0, 2.0, Dimensions(length=2, time=-2))
(+2.0m² s⁻²)

julia> x = AffineDimensions(10.0, 2.0, Dimensions(length=2, time=-2))
(10.0+2.0 m² s⁻²)
```

The last equation looks to be dimensionally invalid.

Can this be polished? How do other packages indicate affine units?